### PR TITLE
More generic grouped grid reduction kernel

### PR DESCRIPTION
**Pull request by naoyam**
This PR generalizes the grouped grid reduction kernel with respect to the number of grouped reductions. The new kernel itself should work with an arbitrary number of inputs, but the underlying data structure, Tuple, still explicitly needs to be specialized for the number of values, which is currently limited to 8. Previously, there's only two-way grouped kernel.

See `FusionGroupAllreduce4`, which groups 8 grid reductions into a single grouped grid reduction.

This PR is meant to allow more aggressive grouping of grid reductions, e.g., grouping across iterations. 

There's still no support for Welford. Fusions with multiple Welford reductions would be unlikely, so horizontal grouping wouldn't be important, but there would be opportunities to group across iterations.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/runtime/fused_reduction.cu [hunk 1]
```diff
@@ -506,34 +879,35 @@ class ParallelReduce {
...
+      const ConstRefTuple<DataTypes...>& inp,
+      VolatilePtrTuple<DataTypes...> global_work_buffer,
+      const LocalTuple<DataTypes...>& init_val,
+      int64_t* global_sync_buffer,
       void* shared_mem,
-      bool read_pred, // Prevent reading from out of bounds memory
-      bool write_pred) { // Prevent from writing out of bounds
+      const LocalTuple<BoolTypes...>& read_preds,
+      const LocalTuple<BoolTypes...>& write_preds,
+      Funcs... funcs) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/runtime/fused_reduction.cu:913]: This is the main change of this PR. Each tuple aggregates a parameter of each grid operations. The number of operations can be as large as 8.

File: torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp [hunk 2]
```diff
@@ -1119,6 +1119,108 @@ TEST_F(NVFuserTest, FusionGroupAllreduce2_CUDA) {
...
 
+// Grouping 3 grid allreduces
+TEST_F(NVFuserTest, FusionGroupAllreduce3_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeSymbolicTensor(1);
+  fusion.addInput(tv0);
+
+  auto tv1 = sum(tv0, {0});
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp:1130]: Could you add some mixed types in the reductions (like one on float vals, one n double vals, and one on int vals)? It looked like that should be supported, right?

**Review by csarofeen — APPROVED** [review]
Didn't do a super in-depth review of all the functions, but overall LGTM. Would like to see an additional test to make sure the tuple's support multiple dtypes across the reductions. Would also just like to see some more comments in the helper functions like the for each functions in the runtime files. Simply reiterating the necessity of the different versions of the functions would be helpful for folks that may need to go through these files in the future. Thanks!

File: torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp:1130]: Yes. Added one more test.

**Comment by naoyam** [comment]
Thanks for the review. Added more comments.

**Pull request merged**